.github/copilot-instructions.md

```diff
@@ -0,0 +1,350 @@
+- Use multi-criteria evaluation (relevance, expertise, depth, clarity, objectivity, completeness)
+- Always include "inferred domain" concept - agents should infer the domain from context
+- Scores use consistent scales (0-10 for single source, 1-5 for comparative scoring)
+- Include clear instructions and examples in prompts
+
+### Dataset Handling
+
+- Use `SamplePairingType` enum for different pairing strategies (relevant, all_existing, embedding_similarity)
+- RAG datasets use triplets: question + source1 + source2
+- Support for ChromaDB vector stores via `langchain_chroma`
+- Use `tqdm` for progress bars in batch operations
+- Checkpoint saves during long processing with `checkpoint_batch_size` parameter
+- CSV export/import for triplet samples with `save_triplets_to_csv()`
+
+### Prompt Templates
+
+The repository uses several prompt templates in `prompts_templates/`:
+
+1. **Source Comparison Prompts** (`rag_verifiers.py`):
+   - `SRC_COMPARE_PROMPT`: Basic comparison without scores
+   - `SRC_COMPARE_PROMPT_WITH_SCORES`: Comparison with 1-5 scores per criterion
+   - `FINAL_VERDICT_PROMPT`: Extract final decision (Source 1/Source 2/Neither)
+   - `SINGLE_SRC_SCORE_PROMPT`: Single source evaluation with 0-10 scores
+   - `CONTEXT_SUFFICIENCY_PROMPT`: Rate context sufficiency 0-5
+
+2. **Generator Prompts** (`rag_generators.py`):
+   - `SUB_TEXT_SPLIT_PROMPT`: Extract relevant vs irrelevant text sections
+   - `QUESTIONS_FROM_2_SOURCES_PROMPT`: Generate questions one source can answer but another cannot
+
+## Common Patterns to Follow
+
+1. **Optional Dependencies**: Wrap imports in try-except blocks with fallback classes for optional features
+   ```python
+   try:
+       from langchain_openai import OpenAIEmbeddings
+   except ImportError:
+       def OpenAIEmbeddings(*args, **kwargs):
+           raise ImportError("...")
+   ```
+
+2. **Batch Processing**: Use `batch_size` parameter and `tqdm` for progress tracking
+
+3. **Empty Text Checking**: Use `_is_empty_text()` helper to check for empty/whitespace/NaN values
+   ```python
+   def _is_empty_text(text: str) -> bool:
+       if not text or text.strip() == "":
+           return True
+       if text.lower() == "nan":
+           return True
+       return False
+   ```
+
+4. **Logging**: Always use module-level `LOGGER = logging.getLogger(__name__)`
+
+5. **Configuration**: Use YAML for model configurations (see `nn/models/model_builder.py`)
+
+6. **ChromaDB Integration**: Use embedding functions with ChromaDB for vector storage
+   ```python
+   self._question_db = Chroma(
+       embedding_function=self._embed_function,
+       persist_directory=os.path.join(dataset_dir, "question_db"),
+   )
+   ```
+
+7. **DataFrame Processing with Checkpoints**:
+   ```python
+   if save_path and checkpoint_batch_size and processed_count % checkpoint_batch_size == 0:
+       result_df.to_csv(save_path, index=False)
+   ```
+
+8. **Fallback Classes for Missing Dependencies**:
```

Comment:
fallback classes are only for agents and anything that using `requirements-agents.txt`